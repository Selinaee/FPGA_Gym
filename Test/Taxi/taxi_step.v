`timescale 1ns / 1ps

module TaxiStep(
    input clk,
    input reset,
    input [2:0] action,
    input [2:0] taxi_row_in,
    input [2:0] taxi_col_in,
    input [2:0] pass_idx_in,
    input [1:0] dest_idx_in,
    output reg [2:0] taxi_row_out,
    output reg [2:0] taxi_col_out,
    output reg [2:0] pass_idx_out,
    output reg [1:0] dest_idx_out,
    output reg [1:0] reward, //3 types of reward: 0(-1), 1(-10), 2(20)
    output reg terminated
);

// 参数定义
parameter MAX_ROW = 4;  // 最大行
parameter MAX_COL = 4;  // 最大列

// reg [4:0] map [0:MAX_ROW] = {
//     5'b01110,  // 行0
//     5'b00000,  // 行1
//     5'b00100,  // 行2
//     5'b00000,  // 行3
//     5'b00010   // 行4
// };
// reg [4:0] map_row_1 = 5'b01110;
// reg [4:0] map_row_2 = 5'b00000;
// reg [4:0] map_row_3 = 5'b00100;
// reg [4:0] map_row_4 = 5'b00000;
// reg [4:0] map_row_5 = 5'b00010;

// reg [24:0] map = {map_row_1[4:0], map_row_2[4:0], map_row_3[4:0], map_row_4[4:0], map_row_5[4:0]};
reg[24:0] map = 25'b0111000000010000000000100;
reg [2:0] pass_row;
reg [2:0] pass_col;
reg [2:0] dest_row;
reg [2:0] dest_col;

always@(*) begin
    case (pass_idx_in)
        0: begin
            pass_row = 0;
            pass_col = 0;
        end
        1: begin
            pass_row = 0;
            pass_col = 4;
        end
        2: begin
            pass_row = 4;
            pass_col = 0;
        end
        3: begin
            pass_row = 4;
            pass_col = 3;
        end
        4: begin
            pass_row = 0;
            pass_col = 0;
        end
    endcase
    case (dest_idx_in)
        0: begin
            dest_row = 0;
            dest_col = 0;
        end
        1: begin
            dest_row = 0;
            dest_col = 4;
        end
        2: begin
            dest_row = 4;
            dest_col = 0;
        end
        3: begin
            dest_row = 4;
            dest_col = 3;
        end
    endcase
end

always @(posedge clk) begin
    if (reset) begin
        // Reset logic
        taxi_row_out <= 0;
        taxi_col_out <= 0;
        pass_idx_out <= 0;
        dest_idx_out <= 0;
        reward <= 0;
        terminated <= 0;
    end else begin
        // Default outputs

        dest_idx_out <= dest_idx_in;
        case (action)
            0: begin// South
                reward <= 0;
                terminated <= 0;
                pass_idx_out <= pass_idx_in;
                if (taxi_row_in < MAX_ROW && map[(5*(taxi_row_in + 1)+(taxi_col_in))] == 0)begin
                    taxi_row_out <= taxi_row_in + 1;
                    taxi_col_out <= taxi_col_in;
                end
                else begin
                    taxi_row_out <= taxi_row_in;
                    taxi_col_out <= taxi_col_in;
                end
            end
            1: begin// North
                reward <= 0;
                terminated <= 0;
                pass_idx_out <= pass_idx_in;
                if (taxi_row_in > 0 && map[(5*(taxi_row_in - 1)+(taxi_col_in))] == 0)begin
                    taxi_row_out <= taxi_row_in - 1;
                    taxi_col_out <= taxi_col_in;
                end
                else begin
                    taxi_row_out <= taxi_row_in;
                    taxi_col_out <= taxi_col_in;
                end
            end
            2: begin// East
                reward <= 0;
                terminated <= 0;
                pass_idx_out <= pass_idx_in;
                if (taxi_col_in < MAX_COL && map[(5*taxi_row_in+(taxi_col_in + 1))] == 0) begin
                    taxi_col_out <= taxi_col_in + 1;
                    taxi_row_out <= taxi_row_in;
                end
                else begin
                    taxi_col_out <= taxi_col_in;
                    taxi_row_out <= taxi_row_in;
                end
            end
            3: begin// West
                reward <= 0;
                terminated <= 0;
                pass_idx_out <= pass_idx_in;
                if (taxi_col_in > 0 && map[(5*taxi_row_in+(taxi_col_in - 1))] == 0) begin
                    taxi_col_out <= taxi_col_in - 1;
                    taxi_row_out <= taxi_row_in;
                end
                else begin
                    taxi_col_out <= taxi_col_in;
                    taxi_row_out <= taxi_row_in;
                end
            end
            4: begin// Pickup
                reward <= 1;  // Assuming pickup is always 1 for simplicity here
                taxi_row_out <= taxi_row_in;
                taxi_col_out <= taxi_col_in;
                terminated <= 0;
                if ((pass_idx_in < 4) && (taxi_row_in == pass_row) && (taxi_col_in == pass_col)) begin
                    pass_idx_out <= 4;
                end
                else begin
                    pass_idx_out <= pass_idx_in;
                end
            end
            5: begin// Dropoff
                taxi_row_out <= taxi_row_in;
                taxi_col_out <= taxi_col_in;
                if ((pass_idx_in == 4) && (taxi_row_in == dest_row) && (taxi_col_in == dest_col)) begin
                    terminated <= 1;
                    reward <= 2;
                    pass_idx_out <= dest_idx_in;
                end 
                else begin
                    terminated <= 0;
                    reward <= 1;
                    pass_idx_out <= pass_idx_in;
                end
            end
        endcase
    end
end

endmodule
