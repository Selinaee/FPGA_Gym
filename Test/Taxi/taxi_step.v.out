#! /c/Source/iverilog-install/bin/vvp
:ivl_version "12.0 (devel)" "(s20150603-1539-g2693dd32b)";
:ivl_delay_selection "TYPICAL";
:vpi_time_precision - 12;
:vpi_module "E:\Software2\iverilog\iverilog\lib\ivl\system.vpi";
:vpi_module "E:\Software2\iverilog\iverilog\lib\ivl\vhdl_sys.vpi";
:vpi_module "E:\Software2\iverilog\iverilog\lib\ivl\vhdl_textio.vpi";
:vpi_module "E:\Software2\iverilog\iverilog\lib\ivl\v2005_math.vpi";
:vpi_module "E:\Software2\iverilog\iverilog\lib\ivl\va_math.vpi";
S_000002501dcf3540 .scope module, "TaxiStep" "TaxiStep" 2 3;
 .timescale -9 -12;
    .port_info 0 /INPUT 1 "clk";
    .port_info 1 /INPUT 1 "reset";
    .port_info 2 /INPUT 3 "action";
    .port_info 3 /INPUT 3 "taxi_row_in";
    .port_info 4 /INPUT 3 "taxi_col_in";
    .port_info 5 /INPUT 3 "pass_idx_in";
    .port_info 6 /INPUT 2 "dest_idx_in";
    .port_info 7 /OUTPUT 3 "taxi_row_out";
    .port_info 8 /OUTPUT 3 "taxi_col_out";
    .port_info 9 /OUTPUT 3 "pass_idx_out";
    .port_info 10 /OUTPUT 2 "dest_idx_out";
    .port_info 11 /OUTPUT 2 "reward";
    .port_info 12 /OUTPUT 1 "terminated";
P_000002501dcfe1c0 .param/l "MAX_COL" 0 2 21, +C4<00000000000000000000000000000100>;
P_000002501dcfe1f8 .param/l "MAX_ROW" 0 2 20, +C4<00000000000000000000000000000100>;
o000002501dd60f98 .functor BUFZ 3, C4<zzz>; HiZ drive
v000002501dcfdf40_0 .net "action", 2 0, o000002501dd60f98;  0 drivers
o000002501dd60fc8 .functor BUFZ 1, C4<z>; HiZ drive
v000002501dcf36d0_0 .net "clk", 0 0, o000002501dd60fc8;  0 drivers
v000002501dcf3770_0 .var "dest_col", 2 0;
o000002501dd61028 .functor BUFZ 2, C4<zz>; HiZ drive
v000002501dcf3810_0 .net "dest_idx_in", 1 0, o000002501dd61028;  0 drivers
v000002501dcf38b0_0 .var "dest_idx_out", 1 0;
v000002501dcf3950_0 .var "dest_row", 2 0;
v000002501ddae030_0 .var "map", 24 0;
v000002501ddae0d0_0 .var "pass_col", 2 0;
o000002501dd61118 .functor BUFZ 3, C4<zzz>; HiZ drive
v000002501ddae170_0 .net "pass_idx_in", 2 0, o000002501dd61118;  0 drivers
v000002501ddae210_0 .var "pass_idx_out", 2 0;
v000002501ddae2b0_0 .var "pass_row", 2 0;
o000002501dd611a8 .functor BUFZ 1, C4<z>; HiZ drive
v000002501ddae350_0 .net "reset", 0 0, o000002501dd611a8;  0 drivers
v000002501ddae3f0_0 .var "reward", 1 0;
o000002501dd61208 .functor BUFZ 3, C4<zzz>; HiZ drive
v000002501ddae490_0 .net "taxi_col_in", 2 0, o000002501dd61208;  0 drivers
v000002501dd41f00_0 .var "taxi_col_out", 2 0;
o000002501dd61268 .functor BUFZ 3, C4<zzz>; HiZ drive
v000002501dd41780_0 .net "taxi_row_in", 2 0, o000002501dd61268;  0 drivers
v000002501dd415a0_0 .var "taxi_row_out", 2 0;
v000002501dd41c80_0 .var "terminated", 0 0;
E_000002501dd39330 .event posedge, v000002501dcf36d0_0;
E_000002501dd39cf0 .event anyedge, v000002501ddae170_0, v000002501dcf3810_0;
    .scope S_000002501dcf3540;
T_0 ;
    %pushi/vec4 14688260, 0, 25;
    %store/vec4 v000002501ddae030_0, 0, 25;
    %end;
    .thread T_0;
    .scope S_000002501dcf3540;
T_1 ;
    %wait E_000002501dd39cf0;
    %load/vec4 v000002501ddae170_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 3;
    %cmp/u;
    %jmp/1 T_1.0, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 3;
    %cmp/u;
    %jmp/1 T_1.1, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 3;
    %cmp/u;
    %jmp/1 T_1.2, 6;
    %dup/vec4;
    %pushi/vec4 3, 0, 3;
    %cmp/u;
    %jmp/1 T_1.3, 6;
    %dup/vec4;
    %pushi/vec4 4, 0, 3;
    %cmp/u;
    %jmp/1 T_1.4, 6;
    %jmp T_1.5;
T_1.0 ;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501ddae2b0_0, 0, 3;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501ddae0d0_0, 0, 3;
    %jmp T_1.5;
T_1.1 ;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501ddae2b0_0, 0, 3;
    %pushi/vec4 4, 0, 3;
    %store/vec4 v000002501ddae0d0_0, 0, 3;
    %jmp T_1.5;
T_1.2 ;
    %pushi/vec4 4, 0, 3;
    %store/vec4 v000002501ddae2b0_0, 0, 3;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501ddae0d0_0, 0, 3;
    %jmp T_1.5;
T_1.3 ;
    %pushi/vec4 4, 0, 3;
    %store/vec4 v000002501ddae2b0_0, 0, 3;
    %pushi/vec4 3, 0, 3;
    %store/vec4 v000002501ddae0d0_0, 0, 3;
    %jmp T_1.5;
T_1.4 ;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501ddae2b0_0, 0, 3;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501ddae0d0_0, 0, 3;
    %jmp T_1.5;
T_1.5 ;
    %pop/vec4 1;
    %load/vec4 v000002501dcf3810_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 2;
    %cmp/u;
    %jmp/1 T_1.6, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 2;
    %cmp/u;
    %jmp/1 T_1.7, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 2;
    %cmp/u;
    %jmp/1 T_1.8, 6;
    %dup/vec4;
    %pushi/vec4 3, 0, 2;
    %cmp/u;
    %jmp/1 T_1.9, 6;
    %jmp T_1.10;
T_1.6 ;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501dcf3950_0, 0, 3;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501dcf3770_0, 0, 3;
    %jmp T_1.10;
T_1.7 ;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501dcf3950_0, 0, 3;
    %pushi/vec4 4, 0, 3;
    %store/vec4 v000002501dcf3770_0, 0, 3;
    %jmp T_1.10;
T_1.8 ;
    %pushi/vec4 4, 0, 3;
    %store/vec4 v000002501dcf3950_0, 0, 3;
    %pushi/vec4 0, 0, 3;
    %store/vec4 v000002501dcf3770_0, 0, 3;
    %jmp T_1.10;
T_1.9 ;
    %pushi/vec4 4, 0, 3;
    %store/vec4 v000002501dcf3950_0, 0, 3;
    %pushi/vec4 3, 0, 3;
    %store/vec4 v000002501dcf3770_0, 0, 3;
    %jmp T_1.10;
T_1.10 ;
    %pop/vec4 1;
    %jmp T_1;
    .thread T_1, $push;
    .scope S_000002501dcf3540;
T_2 ;
    %wait E_000002501dd39330;
    %load/vec4 v000002501ddae350_0;
    %flag_set/vec4 8;
    %jmp/0xz  T_2.0, 8;
    %pushi/vec4 0, 0, 3;
    %assign/vec4 v000002501dd415a0_0, 0;
    %pushi/vec4 0, 0, 3;
    %assign/vec4 v000002501dd41f00_0, 0;
    %pushi/vec4 0, 0, 3;
    %assign/vec4 v000002501ddae210_0, 0;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v000002501dcf38b0_0, 0;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v000002501ddae3f0_0, 0;
    %pushi/vec4 0, 0, 1;
    %assign/vec4 v000002501dd41c80_0, 0;
    %jmp T_2.1;
T_2.0 ;
    %load/vec4 v000002501dcf3810_0;
    %assign/vec4 v000002501dcf38b0_0, 0;
    %load/vec4 v000002501dcfdf40_0;
    %dup/vec4;
    %pushi/vec4 0, 0, 3;
    %cmp/u;
    %jmp/1 T_2.2, 6;
    %dup/vec4;
    %pushi/vec4 1, 0, 3;
    %cmp/u;
    %jmp/1 T_2.3, 6;
    %dup/vec4;
    %pushi/vec4 2, 0, 3;
    %cmp/u;
    %jmp/1 T_2.4, 6;
    %dup/vec4;
    %pushi/vec4 3, 0, 3;
    %cmp/u;
    %jmp/1 T_2.5, 6;
    %dup/vec4;
    %pushi/vec4 4, 0, 3;
    %cmp/u;
    %jmp/1 T_2.6, 6;
    %dup/vec4;
    %pushi/vec4 5, 0, 3;
    %cmp/u;
    %jmp/1 T_2.7, 6;
    %jmp T_2.8;
T_2.2 ;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v000002501ddae3f0_0, 0;
    %pushi/vec4 0, 0, 1;
    %assign/vec4 v000002501dd41c80_0, 0;
    %load/vec4 v000002501ddae170_0;
    %assign/vec4 v000002501ddae210_0, 0;
    %load/vec4 v000002501dd41780_0;
    %pad/u 32;
    %cmpi/u 4, 0, 32;
    %flag_get/vec4 5;
    %jmp/0 T_2.11, 5;
    %load/vec4 v000002501ddae030_0;
    %load/vec4 v000002501dd41780_0;
    %pad/u 32;
    %addi 1, 0, 32;
    %muli 5, 0, 32;
    %load/vec4 v000002501ddae490_0;
    %pad/u 32;
    %add;
    %part/u 1;
    %pad/u 32;
    %pushi/vec4 0, 0, 32;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_2.11;
    %flag_set/vec4 8;
    %jmp/0xz  T_2.9, 8;
    %load/vec4 v000002501dd41780_0;
    %addi 1, 0, 3;
    %assign/vec4 v000002501dd415a0_0, 0;
    %load/vec4 v000002501ddae490_0;
    %assign/vec4 v000002501dd41f00_0, 0;
    %jmp T_2.10;
T_2.9 ;
    %load/vec4 v000002501dd41780_0;
    %assign/vec4 v000002501dd415a0_0, 0;
    %load/vec4 v000002501ddae490_0;
    %assign/vec4 v000002501dd41f00_0, 0;
T_2.10 ;
    %jmp T_2.8;
T_2.3 ;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v000002501ddae3f0_0, 0;
    %pushi/vec4 0, 0, 1;
    %assign/vec4 v000002501dd41c80_0, 0;
    %load/vec4 v000002501ddae170_0;
    %assign/vec4 v000002501ddae210_0, 0;
    %load/vec4 v000002501dd41780_0;
    %pad/u 32;
    %cmpi/u 0, 0, 32;
    %flag_or 5, 4; GT is !LE
    %flag_inv 5;
    %flag_get/vec4 5;
    %jmp/0 T_2.14, 5;
    %load/vec4 v000002501ddae030_0;
    %load/vec4 v000002501dd41780_0;
    %pad/u 32;
    %subi 1, 0, 32;
    %muli 5, 0, 32;
    %load/vec4 v000002501ddae490_0;
    %pad/u 32;
    %add;
    %part/u 1;
    %pad/u 32;
    %pushi/vec4 0, 0, 32;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_2.14;
    %flag_set/vec4 8;
    %jmp/0xz  T_2.12, 8;
    %load/vec4 v000002501dd41780_0;
    %subi 1, 0, 3;
    %assign/vec4 v000002501dd415a0_0, 0;
    %load/vec4 v000002501ddae490_0;
    %assign/vec4 v000002501dd41f00_0, 0;
    %jmp T_2.13;
T_2.12 ;
    %load/vec4 v000002501dd41780_0;
    %assign/vec4 v000002501dd415a0_0, 0;
    %load/vec4 v000002501ddae490_0;
    %assign/vec4 v000002501dd41f00_0, 0;
T_2.13 ;
    %jmp T_2.8;
T_2.4 ;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v000002501ddae3f0_0, 0;
    %pushi/vec4 0, 0, 1;
    %assign/vec4 v000002501dd41c80_0, 0;
    %load/vec4 v000002501ddae170_0;
    %assign/vec4 v000002501ddae210_0, 0;
    %load/vec4 v000002501ddae490_0;
    %pad/u 32;
    %cmpi/u 4, 0, 32;
    %flag_get/vec4 5;
    %jmp/0 T_2.17, 5;
    %load/vec4 v000002501ddae030_0;
    %load/vec4 v000002501dd41780_0;
    %pad/u 32;
    %muli 5, 0, 32;
    %load/vec4 v000002501ddae490_0;
    %pad/u 32;
    %addi 1, 0, 32;
    %add;
    %part/u 1;
    %pad/u 32;
    %pushi/vec4 0, 0, 32;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_2.17;
    %flag_set/vec4 8;
    %jmp/0xz  T_2.15, 8;
    %load/vec4 v000002501ddae490_0;
    %addi 1, 0, 3;
    %assign/vec4 v000002501dd41f00_0, 0;
    %load/vec4 v000002501dd41780_0;
    %assign/vec4 v000002501dd415a0_0, 0;
    %jmp T_2.16;
T_2.15 ;
    %load/vec4 v000002501ddae490_0;
    %assign/vec4 v000002501dd41f00_0, 0;
    %load/vec4 v000002501dd41780_0;
    %assign/vec4 v000002501dd415a0_0, 0;
T_2.16 ;
    %jmp T_2.8;
T_2.5 ;
    %pushi/vec4 0, 0, 2;
    %assign/vec4 v000002501ddae3f0_0, 0;
    %pushi/vec4 0, 0, 1;
    %assign/vec4 v000002501dd41c80_0, 0;
    %load/vec4 v000002501ddae170_0;
    %assign/vec4 v000002501ddae210_0, 0;
    %load/vec4 v000002501ddae490_0;
    %pad/u 32;
    %cmpi/u 0, 0, 32;
    %flag_or 5, 4; GT is !LE
    %flag_inv 5;
    %flag_get/vec4 5;
    %jmp/0 T_2.20, 5;
    %load/vec4 v000002501ddae030_0;
    %load/vec4 v000002501dd41780_0;
    %pad/u 32;
    %muli 5, 0, 32;
    %load/vec4 v000002501ddae490_0;
    %pad/u 32;
    %subi 1, 0, 32;
    %add;
    %part/u 1;
    %pad/u 32;
    %pushi/vec4 0, 0, 32;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_2.20;
    %flag_set/vec4 8;
    %jmp/0xz  T_2.18, 8;
    %load/vec4 v000002501ddae490_0;
    %subi 1, 0, 3;
    %assign/vec4 v000002501dd41f00_0, 0;
    %load/vec4 v000002501dd41780_0;
    %assign/vec4 v000002501dd415a0_0, 0;
    %jmp T_2.19;
T_2.18 ;
    %load/vec4 v000002501ddae490_0;
    %assign/vec4 v000002501dd41f00_0, 0;
    %load/vec4 v000002501dd41780_0;
    %assign/vec4 v000002501dd415a0_0, 0;
T_2.19 ;
    %jmp T_2.8;
T_2.6 ;
    %pushi/vec4 1, 0, 2;
    %assign/vec4 v000002501ddae3f0_0, 0;
    %load/vec4 v000002501dd41780_0;
    %assign/vec4 v000002501dd415a0_0, 0;
    %load/vec4 v000002501ddae490_0;
    %assign/vec4 v000002501dd41f00_0, 0;
    %pushi/vec4 0, 0, 1;
    %assign/vec4 v000002501dd41c80_0, 0;
    %load/vec4 v000002501ddae170_0;
    %pad/u 32;
    %cmpi/u 4, 0, 32;
    %flag_get/vec4 5;
    %jmp/0 T_2.24, 5;
    %load/vec4 v000002501dd41780_0;
    %load/vec4 v000002501ddae2b0_0;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_2.24;
    %flag_set/vec4 9;
    %flag_get/vec4 9;
    %jmp/0 T_2.23, 9;
    %load/vec4 v000002501ddae490_0;
    %load/vec4 v000002501ddae0d0_0;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_2.23;
    %flag_set/vec4 8;
    %jmp/0xz  T_2.21, 8;
    %pushi/vec4 4, 0, 3;
    %assign/vec4 v000002501ddae210_0, 0;
    %jmp T_2.22;
T_2.21 ;
    %load/vec4 v000002501ddae170_0;
    %assign/vec4 v000002501ddae210_0, 0;
T_2.22 ;
    %jmp T_2.8;
T_2.7 ;
    %load/vec4 v000002501dd41780_0;
    %assign/vec4 v000002501dd415a0_0, 0;
    %load/vec4 v000002501ddae490_0;
    %assign/vec4 v000002501dd41f00_0, 0;
    %load/vec4 v000002501dcf3810_0;
    %pad/u 3;
    %assign/vec4 v000002501ddae210_0, 0;
    %load/vec4 v000002501ddae170_0;
    %pad/u 32;
    %cmpi/e 4, 0, 32;
    %flag_get/vec4 4;
    %jmp/0 T_2.28, 4;
    %load/vec4 v000002501dd41780_0;
    %load/vec4 v000002501dcf3950_0;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_2.28;
    %flag_set/vec4 9;
    %flag_get/vec4 9;
    %jmp/0 T_2.27, 9;
    %load/vec4 v000002501ddae490_0;
    %load/vec4 v000002501dcf3770_0;
    %cmp/e;
    %flag_get/vec4 4;
    %and;
T_2.27;
    %flag_set/vec4 8;
    %jmp/0xz  T_2.25, 8;
    %pushi/vec4 1, 0, 1;
    %assign/vec4 v000002501dd41c80_0, 0;
    %pushi/vec4 2, 0, 2;
    %assign/vec4 v000002501ddae3f0_0, 0;
    %jmp T_2.26;
T_2.25 ;
    %pushi/vec4 0, 0, 1;
    %assign/vec4 v000002501dd41c80_0, 0;
    %pushi/vec4 1, 0, 2;
    %assign/vec4 v000002501ddae3f0_0, 0;
T_2.26 ;
    %jmp T_2.8;
T_2.8 ;
    %pop/vec4 1;
T_2.1 ;
    %jmp T_2;
    .thread T_2;
# The file index is used to find the file name in the following table.
:file_names 3;
    "N/A";
    "<interactive>";
    "taxi_step.v";
